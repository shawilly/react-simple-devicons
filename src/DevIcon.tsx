import { type CSSProperties, useCallback, useEffect, useState } from "react";
import { Loader } from "./components/loader";
import { type Scale, dimensions } from "./lib/constants/size";
import { deviconSvgUrl } from "./lib/devicon-svg-url";

const commonError =
	" Double check icon availability here: https://devicon.dev/";

export type DevIconProps = {
	color?: string;
	icon?: string;
	scale?: Scale;
	style?:
		| "line"
		| "line-woodmark"
		| "original"
		| "original-woodmark"
		| "plain"
		| "plain-woodmark";
};

export function DevIcon(props: DevIconProps) {
	const icon = props?.icon || "devicon";
	const iconStyle = props?.style || "plain";

	const [svgContent, setSvgContent] = useState<string | null>(null);
	const [dimension, setDimension] = useState<number>();
	const [isLoading, setIsLoading] = useState(true);

	/**
	 * Throw error with provided message, set isLoading to false and svgContent to null.
	 */
	const throwError = (message: string): void => {
		setIsLoading(false);
		setSvgContent(null);

		throw new Error(message + commonError);
	};

	const handleError = useCallback(throwError, []);

	useEffect((): void => {
		// Set scale
		if (props?.scale && dimensions[props.scale]) {
			setDimension(dimensions[props?.scale]);
		} else {
			setDimension(14);
		}

		// You can't change color of original style
		if (props?.style?.includes("original") && props?.color) {
			handleError(
				"Can't set color of original style. Try line(-woodmark) or plain(-woodmark) styles.",
			);
		}
	}, [props?.color, props?.scale, props?.style, handleError]);

	useEffect((): void => {
		setIsLoading(true);
		const fetchSvg = async (): Promise<void> => {
			try {
				const response = await fetch(deviconSvgUrl(icon, iconStyle));

				if (!response.ok) {
					handleError(
						`Fetch error: Unable to fetch icon '${icon}' with style '${iconStyle}'. HTTP status: ${response.status}`,
					);
				}

				const text = await response.text();

				if (!text) {
					handleError(
						`No svg text found for icon '${icon}' with style '${iconStyle}'. HTTP status: ${response.status}`,
					);
				}

				setSvgContent(text);
				setIsLoading(false);
			} catch (e) {
				handleError(`Error fetching the icon: ${(e as Error).message}`);
			}
		};

		fetchSvg(); // Directly call fetchSvg without a try-catch here
	}, [icon, iconStyle, handleError]);

	if (!(svgContent || isLoading)) {
		throwError(`Icon '${icon}' with style '${iconStyle}' not found.`);
	}

	const style: CSSProperties = {
		width: dimension,
		height: dimension,
		color: props.color || "inherit", // Use inherit if no color is provided
	};

	return isLoading || !svgContent ? (
		<Loader />
	) : (
		<div
			id={`dev-icon-${icon}-${iconStyle}`}
			data-testid="dev-icon"
			style={style}
			// Replace all color values to the provided color
			dangerouslySetInnerHTML={{
				__html: svgContent.replace(
					/fill="[^"]*"/g,
					`fill="${props.color || "currentColor"}"`,
				),
			}}
		/>
	);
}
