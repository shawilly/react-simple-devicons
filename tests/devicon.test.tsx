import { render, screen } from "@testing-library/react";
import { expect, it, vi } from "vitest";
import { DevIcon, type DevIconProps } from "../src";

const props: DevIconProps = {
	color: "black",
	icon: "typescript",
	scale: "lg",
	style: "plain",
};

// Create a mock implementation for fetch
global.fetch = vi.fn(() =>
	Promise.resolve({
		ok: true,
		status: 200,
		statusText: "OK",
		headers: new Headers(),
		url: "http://example.com/icon.svg",
		text: () => Promise.resolve('<svg><path d="M10 20"></path></svg>'),
	}),
) as unknown as typeof global.fetch;

it("fails to render original style with color", () => {
	expect(() => {
		render(<DevIcon {...props} style="original" color="red" />);
	}).toThrowError();
});

it("loads and renders the icon SVG content successfully", async () => {
	render(<DevIcon {...props} />);

	// Use the method of your choice to find the SVG element
	const svgElement = await screen.findByTestId("dev-icon");
	expect(svgElement).toBeInTheDocument();
});

it("matches the snapshot", () => {
	const { asFragment } = render(<DevIcon {...props} />);
	expect(asFragment()).toMatchSnapshot();
});
