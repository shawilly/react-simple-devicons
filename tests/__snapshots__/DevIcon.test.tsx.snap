// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="loader"
    data-testid="loader"
  />
</DocumentFragment>
`;
